File: backend/services/card_service.py
import json
import random
from pathlib import Path
from typing import List, Dict
from schemas.game_schema import Card, QuestionType

# Load cards from JSON file
CARDS_FILE = Path(__file__).parent.parent / "data" / "cards.json"


def load_cards() -> List[Dict]:
    """Load all card definitions from JSON file"""
    with open(CARDS_FILE, "r") as f:
        return json.load(f)


def get_allowed_difficulty(game_size: int) -> str:
    """
    Returns the single allowed difficulty level based on game size.

    game_size 3 -> "Small"
    game_size 4 -> "Medium"
    game_size 5 -> "Large"
    """
    if game_size == 3:
        return "Small"
    elif game_size == 4:
        return "Medium"
    else:  # game_size == 5
        return "Large"


def generate_random_cards(count: int, game_size: int = 5) -> List[Card]:
    """
    Generate random cards from the card pool, filtered by game size (difficulty level).
    Only includes cards that have non-zero values in the allowed difficulty column,
    OR special cards (Discard types).

    Args:
        count: Number of cards to generate
        game_size: Game difficulty level (3=Small, 4=Medium, 5=Large)

    Returns:
        List of Card objects
    """
    all_cards = load_cards()
    allowed_difficulty = get_allowed_difficulty(game_size)

    # Create a weighted pool based on 'cards' count, filtered by difficulty
    card_pool = []
    for card_def in all_cards:
        # Include if it's an Action card OR has value in allowed difficulty
        is_action_card = card_def["Type"] == "Action"
        has_allowed_difficulty = card_def.get(allowed_difficulty, 0) > 0

        if is_action_card or has_allowed_difficulty:
            # Add each card type 'cards' times to the pool
            for _ in range(card_def["cards"]):
                card_pool.append(card_def)

    # Randomly select 'count' cards from the pool
    selected_cards = random.sample(card_pool, min(count, len(card_pool)))

    # Convert to Card schema objects
    return [Card(**card) for card in selected_cards]


def get_draw_count(question_type: QuestionType) -> tuple[int, int]:
    """
    Get the draw count and pick count for a question type.
    Returns: (draw_count, pick_count)
    """
    draw_pick_mapping = {
        QuestionType.MATCHING: (3, 1),
        QuestionType.MEASURING: (3, 1),
        QuestionType.THERMOMETER: (2, 1),
        QuestionType.RADAR: (2, 1),
        QuestionType.TENTACLES: (4, 2),
        QuestionType.PHOTOS: (1, 1),
    }
    return draw_pick_mapping.get(question_type, (0, 0))

File: api/services/cardService.js
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

let cardsData = null;

export function loadCards() {
  if (!cardsData) {
    const cardsPath = join(__dirname, '..', 'cards.json');
    cardsData = JSON.parse(readFileSync(cardsPath, 'utf-8'));
  }
  return cardsData;
}

export function getAllowedDifficulty(gameSize) {
  const difficultyMap = {
    3: 'Small',
    4: 'Medium',
    5: 'Large',
  };
  return difficultyMap[gameSize] || 'Large';
}

export function generateRandomCards(count, gameSize) {
  const cards = loadCards();
  const difficulty = getAllowedDifficulty(gameSize);

  // Filter cards that are valid for this difficulty
  const validCards = cards.filter(card => {
    // Action and Curse cards are always included
    if (card.Type === 'Action' || card.Type === 'Curse') {
      return true;
    }
    // Time Bonus cards need a value > 0 for this difficulty
    return card[difficulty] > 0;
  });

  // Create weighted pool based on card.cards frequency
  const weightedPool = [];
  validCards.forEach(card => {
    for (let i = 0; i < card.cards; i++) {
      weightedPool.push(card);
    }
  });

  // Sample random cards with replacement
  const drawnCards = [];
  for (let i = 0; i < count; i++) {
    const randomIndex = Math.floor(Math.random() * weightedPool.length);
    drawnCards.push({ ...weightedPool[randomIndex] });
  }

  return drawnCards;
}

// Build a full shuffled deck based on game size
export function buildDeck(gameSize) {
  const cards = loadCards();
  const difficulty = getAllowedDifficulty(gameSize);
  const deck = [];

  for (const card of cards) {
    const isValid =
      card.Type === 'Action' ||
      card.Type === 'Curse' ||
      card[difficulty] > 0;

    if (isValid) {
      for (let i = 0; i < card.cards; i++) {
        deck.push({ ...card });
      }
    }
  }

  // Shuffle the deck
  for (let i = deck.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [deck[i], deck[j]] = [deck[j], deck[i]];
  }

  return deck;
}

// Draw cards from deck for a question type draw.
// Clears any previously pending cards (returns them to pool),
// then marks `count` new cards as pending.
export function drawFromDeck(deck, count) {
  // Clear existing pending markers so previous unconfirmed draws return to pool
  const cleanDeck = deck.map(card => {
    if (card._pending) {
      const { _pending, ...rest } = card;
      return rest;
    }
    return card;
  });

  if (cleanDeck.length < count) {
    const error = new Error('Deck exhausted. Reset is required to continue drawing.');
    error.code = 'DECK_EXHAUSTED';
    throw error;
  }

  // Shuffle indices and pick `count` distinct ones
  const indices = Array.from({ length: cleanDeck.length }, (_, i) => i);
  for (let i = indices.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [indices[i], indices[j]] = [indices[j], indices[i]];
  }
  const pickedSet = new Set(indices.slice(0, count));

  const updatedDeck = cleanDeck.map((card, idx) =>
    pickedSet.has(idx) ? { ...card, _pending: true } : card
  );

  const drawn = [...pickedSet].map(idx => {
    const { _pending, ...cardWithoutPending } = updatedDeck[idx];
    return cardWithoutPending;
  });

  return { drawn, updatedDeck };
}

// Immediately consume `count` cards from the deck (for Discard X Draw Y).
// Removes them from the deck permanently without a pending step.
export function consumeFromDeck(deck, count) {
  // Clear any pending markers first
  const cleanDeck = deck.map(card => {
    if (card._pending) {
      const { _pending, ...rest } = card;
      return rest;
    }
    return card;
  });

  if (cleanDeck.length < count) {
    const error = new Error('Deck exhausted. Reset is required to continue drawing.');
    error.code = 'DECK_EXHAUSTED';
    throw error;
  }

  // Shuffle and pick
  const indices = Array.from({ length: cleanDeck.length }, (_, i) => i);
  for (let i = indices.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [indices[i], indices[j]] = [indices[j], indices[i]];
  }
  const pickedSet = new Set(indices.slice(0, count));

  const drawn = [...pickedSet].map(idx => cleanDeck[idx]);
  const updatedDeck = cleanDeck.filter((_, idx) => !pickedSet.has(idx));

  return { drawn, updatedDeck };
}

export function getDrawCount(questionType) {
  const drawMap = {
    MATCHING: { draw: 3, pick: 1 },
    MEASURING: { draw: 3, pick: 1 },
    THERMOMETER: { draw: 2, pick: 1 },
    RADAR: { draw: 2, pick: 1 },
    TENTACLES: { draw: 4, pick: 2 },
    PHOTOS: { draw: 1, pick: 1 },
  };
  return drawMap[questionType] || { draw: 3, pick: 1 };
}
